cpp/src/io/comp/nvcomp_adapter.cpp

```diff
@@ -321,40 +324,68 @@ static void batched_compress_async(compression_type compression,
   CUDF_EXPECTS(nvcomp_status == nvcompStatus_t::nvcompSuccess, "Error in compression");
 }
 
-bool is_aligned(void const* ptr, std::uintptr_t alignment) noexcept
+constexpr bool is_aligned(void const* ptr, std::uintptr_t alignment) noexcept
 {
   return (reinterpret_cast<std::uintptr_t>(ptr) % alignment) == 0;
 }
 
+// since some compressors have extreme temp memory requirements, scale back the number
+// of chunks to process until the request for temp memory succeeds.
+std::pair<rmm::device_buffer, size_t> compress_temp_buffer(compression_type compression,
+                                                           size_t num_chunks,
+                                                           size_t max_uncomp_chunk_size,
+                                                           rmm::cuda_stream_view stream)
+{
+  auto scaled_num_chunks = num_chunks;
+  size_t scale           = 1;
+  while (scaled_num_chunks > 0) {
```

Comment:
This is a potential perf regression for Spark, since they have an option to spill and make more memory available. If we end up compressing a single chunk at a time, performance could be a lot worse than spilling. CC @revans2 
Maybe we should bail at `sqrt(num_chunks)`? It makes the minimum number of concurrent chunks the same as the number of retries ¯\_(ツ)_/¯

Reply:
I've been all over the place on this and don't have my heart set on anything. I'm wondering, though, if the spill would happen on file writing, or just during query execution?  Or does Spark spill to parquet files?

Reply:
After running some tests, I'm wondering if this should go back to a geometric scaling rather than linear.  In one case It tried 41 times until succeeding 😮

Reply:
wow! What was the initial number of chunks?
I'm okay with geometric :+1:

Reply:
> wow! What was the initial number of chunks?

Around 2400.  Oddly, I get worse execution time with geometric scaling since it overshoots most of the time, resulting in more compression batches than necessary.

Reply:
Hmm. I assume that's doubling at each step. Have you tried geometric with lower ratio?

Reply:
`sqrt(2)` worked pretty well.  Got 41 iterations down to 11, and still got good performance.  How's that for a compromise?

Reply:
Sounds good!